renku/core/utils/templates.py

```diff
@@ -0,0 +1,871 @@
+                immutable_template_files=immutable_template_files,
+            )
+        else:
+            raise errors.ParameterError(f"Invalid template render action: {action}")
+
+        checksum = hash_file(render_path) or hash_file(path_in_project)
+        if checksum is not None:
+            new_checksums[relative_path] = checksum
+
+    return render_base, new_checksums
+
+
+def _render_file_for_initialize(
+    render_path: Path,
+    path_in_project: Path,
+    relative_path: str,
+    rendered_content: str,
+):
+    def should_append(path: str):
+        return path.lower() in TEMPLATE_INIT_APPEND_FILES
+
+    def should_keep(path: str):
+        return path.lower() in TEMPLATE_KEEP_FILES
+
+    if not path_in_project.exists():
+        communication.echo(f"Initializing {relative_path} ...")
+        render_path.write_text(rendered_content)
+    elif should_append(relative_path):
+        communication.echo(f"Appending to {relative_path} ...")
+        render_path.write_text(path_in_project.read_text() + "\n" + rendered_content)
+    elif should_keep(relative_path):
+        communication.echo(f"Keeping file {relative_path} ...")
+    else:
+        communication.echo(f"Overwriting {relative_path} ...")
+        render_path.write_text(rendered_content)
+
+
+def _render_file_for_set(
+    render_path: Path,
+    path_in_project: Path,
+    relative_path: str,
+    rendered_content: str,
+    interactive: bool,
+):
+    """Decide what to do with files in a project when setting a template."""
+
+    def should_keep(path: str):
+        return path.lower() in TEMPLATE_KEEP_FILES
+
+    file_exists = path_in_project.exists()
+
+    if not file_exists:
+        communication.echo(f"Initializing {relative_path} ...")
+        render_path.write_text(rendered_content)
+    elif interactive:
+        if communication.confirm(f"Overwrite {relative_path}?", default=True):
+            render_path.write_text(rendered_content)
+    elif should_keep(relative_path):
+        communication.echo(f"Keeping file {relative_path} ...")
+    else:
+        communication.echo(f"Overwriting {relative_path} ...")
+        render_path.write_text(rendered_content)
+
+
+def _render_file_for_update(
+    render_path: Path,
+    path_in_project: Path,
+    relative_path: str,
+    rendered_content: str,
+    checksums: Optional[Dict[str, str]],
+    interactive: bool,
+    immutable_template_files: List[str],
+):
+    """Decide what to do with files in a project when updating a template."""
+    current_hash = hash_file(path_in_project)
+    local_changes = current_hash != checksums.get(relative_path)
+    new_hash = hash_str(rendered_content)
+    remote_changes = new_hash != checksums.get(relative_path)
+    file_exists = path_in_project.exists()
+    file_deleted = not file_exists and relative_path in checksums
+
+    if not file_exists and not file_deleted:
+        communication.echo(f"Initializing {relative_path} ...")
```

Comment:
We recently discussed that we use `"my text '{value}'"` form in messages, I think we should also apply this here (I didn't explicitely look for this in the rest of the PR, there might be other places)